keps/sig-cli/3517-kubectl-client-side-apply-migration/README.md

```diff
@@ -0,0 +1,1113 @@
+adoption for server-side-apply is blocked for many users.
+
+1. The primary goal of this KEP is to provide a pathway for users blocked by this issue to adopt SSA.
+2. Provide an idempotent migration to convert large collections of resources to server-side apply, especially for automatic systems
+3. Keep the number of requests to the apiserver limited on the nominal server-side apply path (no additional writes)
+
+
+### Non-Goals
+
+1. Provide a server-side fix to this problem (see alternatives considered)
+2. Deprecate the `kubectl/last-applied-configuration` annotation
+
+## Proposal
+
+Provide a transparent migration path of client-side-apply field ownership to
+being owned by server-side-apply.
+
+```shell
+kubectl apply -f <path/to/file> --server-side
+```
+
+Today when using server-side-apply, kubectl unconditionally does an initial fetch
+of the object. This is an acknowledged anti-pattern of SSA, and will likely be changed
+in the future. Since it is pre-existing behavior that does not do much harm, the SIG has
+come to consensus that the behavior can remain for a few more releases until client-side-apply
+is out of vogue, especially since SSA is now GA.
+
+The initially fetched object will be tested to check if the migration should be
+performed. If not, SSA proceeds as normal. If the migration is required, then
+a `UPDATE` or JSON PATCH is sent prior to the normal SSA operation. This is a technical requirement:
+managed fields may not be modified within an SSA operation.
+
+### Example Usage
+
+The usage should be transparent to the user, so they don't have to learn any new
+interface. Users of server-side-apply will have their fields owned by client-side-apply
+transparently migrated for them.
+
+```shell
+# Migrate fields owned by kubectl-client-side-apply to now be owned by `myssamanager`
+kubectl apply -f <path/to/file> --server-side --fieldmanager=myssamanager
+```
+
+Before continuing with the apply operation, kubectl will give ownership to all
+fields previously managed through client-side-apply, to server-side-apply.
+And drop client-side-apply's claim on the field.
+
+If client-side-apply does not own any fields, then the behavior is to no-op and
+continue with the `apply` invocation.
+
+### Transferring Field Ownership
+
+The fields previously managed by kubectl-client-side-apply before this operation
+will afterwoards be managed by whatever field manager was provided to the command
+line via the `--fieldmanager` flag.
+
+If a user wishes any specific fields to then be owned by other field managers piecemeal,
+then the user may do so normally using documentation provided about transferring
+field ownership.
+
+### Risks and Mitigations
+
+#### Operation reversion
+
+If the user reverts to using client-side-apply again, this operation's effect
+will be reverted. Since the future is for everyone to use server-side-apply only
```

Comment:
If ever a client-side-apply operation is undertaken, the apiserver has logic to recreate the managedFields for the `Update` operation. All the fields provided by the `Update` will then be included in the new managed fields. So this reversion would happen automatically until all `kubectl` clients have migrated to use SSA.

Reply:
Am I understanding correctly that this KEP isn't proposing any changes to the reversion flow then? IIUC it can still result in a situation where SSA gets left with some managedFields after reversion, but that doesn't really matter in practice because CSA necessarily believes it owns those fields at takeover (because its annotation is a mirror of them) and it is unaware of the conflict since it ignores managedFields entirely. Is that right?

Reply:
Correct, the reversion flow remains unchanged. 

Every time CSA is used after SSA, the annotation will have been up to date since apiserver maintains it for SSA. The operation proceeds as normal and adds back a claim onto managed fields. As you correctly mention, any fields owned by the SSA manager will not affect the operation done by CSA. This is the behavior today and not a change in this KEP.

Every time SSA is used after CSA, this KEP will cause `kubectl` to again 'fix' the managed fields so that the operation proceeds as expected to the user. This is new behavior from the KEP.

Eventually a steady state is reached where all usages of client-side-apply are removed and the managed fields remain stable.